[Draft] CI: flaky Windows test rel-csv-split (run 24850653533)

## Summary

The failing job is **ci / tests / windows-latest** in workflow **Release Candidate** (not the Docker change itself).

- **Run:** https://github.com/abhigyanpatwari/GitNexus/actions/runs/24850653533/job/72749889772
- **Head SHA:** \c9246cfd31a8b2e828283b04ce56806c6f94168\ — merge **#1029** (\ci(docker): mirror signed images...\)
- **That commit only touched:** \.github/workflows/docker.yml\, \.env.example\, \README.md\, \CONTRIBUTING.md\, \deploy/kubernetes/cluster-image-policy.yaml\ — **no** \lbug-adapter\ / \
el-csv-split\ code.

## Root cause

**Not a regression introduced by #1029.** The red job failed in \	est/unit/rel-csv-split.test.ts\, test **\destroys all streams when one errors (no lingering FDs)\**:

- \AssertionError: expected +0 to be 1\ at the check \xpect(streams.length).toBe(1)\ after a **20 ms** \setTimeout\.
- On a busy **windows-latest** runner, readline/async scheduling can lag so the first mock stream is not created within 20 ms → assertion fails → test tears down while work is still in flight → **\ENOTEMPTY\** on temp dir cleanup (secondary symptom).

## Fix (proposed)

Replace fixed sleeps with **\wait expect.poll(...)\** (Vitest 4) so the test waits until \streams.length\ is actually 1, then ≥2, with a bounded timeout — same intent, deterministic on slow CI.

---

*This issue was opened as a draft investigation note; close or retitle once a PR with the test hardening is merged.*

--- a/gitnexus/test/unit/rel-csv-split.test.ts
+++ b/gitnexus/test/unit/rel-csv-split.test.ts
@@ -247,14 +247,14 @@ describe('splitRelCsvByLabelPair', () => {
       mockFactory(streams, { blocked: true }),
     );
 
-    // The first pair stream is created immediately and blocks on its header
-    // write. Unblock it once so the loop advances and creates the second
-    // pair stream (also blocked). Now both streams exist — trigger the error.
-    await new Promise((r) => setTimeout(r, 20));
-    expect(streams.length).toBe(1);
+    // The first pair stream is created once readline has delivered a data row;
+    // fixed sleeps race on slow Windows CI (streams.length still 0 → ENOTEMPTY
+    // in afterEach). Poll until the harness reaches the intended state.
+    await expect.poll(() => streams.length, { interval: 10, timeout: 10_000 }).toBe(1);
     streams[0].unblock();
-    await new Promise((r) => setTimeout(r, 20));
-    expect(streams.length).toBeGreaterThanOrEqual(2);
+    await expect
+      .poll(() => streams.length, { interval: 10, timeout: 10_000 })
+      .toBeGreaterThanOrEqual(2);
     streams[0].triggerError(new Error('EMFILE'));
 
     await expect(promise).rejects.toThrow('EMFILE');

--- a/gitnexus/test/unit/rel-csv-split.test.ts
+++ b/gitnexus/test/unit/rel-csv-split.test.ts
@@ -106,6 +106,8 @@ function writeCsv(lines: string[]): string {
 // ---------------------------------------------------------------------------
 describe('splitRelCsvByLabelPair', () => {
   const validTables = new Set(['Function', 'Class', 'File', 'Method']);
+  /** Bounded poll for readline + split loop to reach an observable milestone (DoD §2.7). */
+  const pollOpts = { interval: 10, timeout: 10_000 } as const;
 
   it('splits lines into per-pair files with correct row counts', async () => {
     const csvPath = writeCsv([
@@ -197,8 +199,8 @@ describe('splitRelCsvByLabelPair', () => {
       mockFactory(streams, { blocked: true }),
     );
 
-    // Give readline time to buffer and fire lines
-    await new Promise((r) => setTimeout(r, 50));
+    // All rows share Function|Class — one stream, blocked on header or row drain
+    await expect.poll(() => streams.length, pollOpts).toBe(1);
 
     // Unblock all streams so the Promise can resolve
     for (const ws of streams) ws.unblock();
@@ -222,9 +224,7 @@ describe('splitRelCsvByLabelPair', () => {
       mockFactory(streams, { blocked: true }),
     );
 
-    // Wait for readline to process, then error while paused on drain
-    await new Promise((r) => setTimeout(r, 50));
-    expect(streams.length).toBe(1);
+    await expect.poll(() => streams.length, pollOpts).toBe(1);
     streams[0].triggerError(new Error('disk full'));
 
     await expect(promise).rejects.toThrow('disk full');
@@ -247,14 +247,12 @@ describe('splitRelCsvByLabelPair', () => {
       mockFactory(streams, { blocked: true }),
     );
 
-    // The first pair stream is created once readline has delivered a data row;
-    // fixed sleeps race on slow Windows CI (streams.length still 0 → ENOTEMPTY
-    // in afterEach). Poll until the harness reaches the intended state.
-    await expect.poll(() => streams.length, { interval: 10, timeout: 10_000 }).toBe(1);
+    // First pair stream once readline delivered a row; poll avoids Windows CI races.
+    await expect.poll(() => streams.length, pollOpts).toBe(1);
     streams[0].unblock();
     // Exactly two pair keys before the third CSV row: Function|Class then
     // File|Method; the loop is blocked on the second stream's header drain.
-    await expect.poll(() => streams.length, { interval: 10, timeout: 10_000 }).toBe(2);
+    await expect.poll(() => streams.length, pollOpts).toBe(2);
     streams[0].triggerError(new Error('EMFILE'));
 
     await expect(promise).rejects.toThrow('EMFILE');

--- a/gitnexus/test/unit/rel-csv-split.test.ts
+++ b/gitnexus/test/unit/rel-csv-split.test.ts
@@ -224,7 +224,7 @@ describe('splitRelCsvByLabelPair', () => {
 
     // Wait for readline to process, then error while paused on drain
     await new Promise((r) => setTimeout(r, 50));
-    expect(streams.length).toBeGreaterThan(0);
+    expect(streams.length).toBe(1);
     streams[0].triggerError(new Error('disk full'));
 
     await expect(promise).rejects.toThrow('disk full');
@@ -252,9 +252,9 @@ describe('splitRelCsvByLabelPair', () => {
     // in afterEach). Poll until the harness reaches the intended state.
     await expect.poll(() => streams.length, { interval: 10, timeout: 10_000 }).toBe(1);
     streams[0].unblock();
-    await expect
-      .poll(() => streams.length, { interval: 10, timeout: 10_000 })
-      .toBeGreaterThanOrEqual(2);
+    // Exactly two pair keys before the third CSV row: Function|Class then
+    // File|Method; the loop is blocked on the second stream's header drain.
+    await expect.poll(() => streams.length, { interval: 10, timeout: 10_000 }).toBe(2);
     streams[0].triggerError(new Error('EMFILE'));
 
     await expect(promise).rejects.toThrow('EMFILE');

